### PR TITLE
Match bin data structure

### PR DESCRIPTION
**Pull request by mmore500**
Let's make a general data structure that we can store value-tag pairs in and then use a configurable match metric (e.g., hamming distance) and selector (e.g., all matches above a threshold or roulette picking) to access values based on a tag query.

### CODE REVIEW

File: source/tools/MatchBin.h [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+#include <stdexcept>
+
+#include "../base/assert.h"
+#include "../base/vector.h"
+#include "../tools/IndexMap.h"
+
+namespace emp {
+
+  template<size_t width>
+  struct HammingDistance{
```
Comment (mmore500) [inline source/tools/MatchBin.h]: Looking at how the `std::hash<emp::BitSet<N>>` is implemented, maybe after all it would be a better idea to not have a static `Metric` function and do a non-static `operator()` function. I assume that the people who made that design decision for `std::hash` stuff know best so copying them as closely as possible in this case would probably be a good idea.

File: source/tools/MatchBin.h [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    static double Metric(BitSet<width> a, BitSet<width> b){return (a^b).CountOnes();}
+  };
+
+  struct Difference{
+    static double Metric(int a, int b){ return (double)abs(a-b);}
+  };
+
+  template<int numerator, int denominator>
```
Comment (mmore500) [inline source/tools/MatchBin.h]: It should be functionally equivalent to what we're doing here, but I think using [`std::ratio`](http://www.cplusplus.com/reference/ratio/ratio/) would better express our intent/be more standard.

File: source/tools/MatchBin.h [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  struct ThreshSelector{
```
Comment (mmore500) [inline source/tools/MatchBin.h]: Same comment as on line 29.

File: source/tools/MatchBin.h [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        );
+        while(i < uids.size() && i < n && scores.at(uids[i]) < thresh) ++i;   
+      }
+      return emp::vector<size_t>(uids.begin(), uids.begin()+i);
+    }
+  };
+  
+  /// Factory function that generates a selector function that chooses
+  /// probabilisticly based on match quality with replacement.
+  struct RouletteSelector{
```
Comment (mmore500) [inline source/tools/MatchBin.h]: Same comment as on line 29

File: source/tools/MatchBin.h [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+  /// functionality is also provided, allowing dynamically adjustment of match
+  /// strength to a particular item (i.e., making all matches stronger/weaker).
+  /// A unique identifier is generated upon tag/item placement in the container.
+  /// This unique identifier can be used to view or edit the stored items and
+  /// their corresponding tags. Tag-based lookups return a list of matched
+  /// unique identifiers.
+  template <typename Val, typename Tag, typename Metric, typename Selector> class MatchBin {
+  private:
+
+    //const std::function<double(Tag, Tag)> metric;
```
Comment (mmore500) [inline source/tools/MatchBin.h]: When we go through at the end to update the documentation and in-code comments, let's delete these.

File: tests/test_config.cc [hunk 6]
```diff
@@ -40,7 +40,7 @@ TEST_CASE("Test config", "[config]"){
   // old ArgManager in cl namespace
   {
 
-    emp::vector<std::string> arguments = {"--dir", "/some_path"};
```
Comment (mmore500) [inline tests/test_config.cc]: I don't know why these unrelated changes are here... I think these have already been merged in on PR #239 . We'll need to double check and maybe fix the Git history.

Comment (mmore500) [reply tests/test_config.cc]: (my bad on this one)

File: source/tools/BitSet.h [hunk 7]
```diff
@@ -610,11 +618,28 @@ namespace emp {
 
 }
 
-template <size_t NUM_BITS> std::ostream & operator<<(std::ostream & out, const emp::BitSet<NUM_BITS> & _bit_set) {
-  _bit_set.Print(out);
-  return out;
+/// For hashing BitSets
+namespace std
```
[the remark this answers is not in the document]
Comment (mercere99) [reply source/tools/BitSet.h]: Technically, this is the "correct" way to implement various features (including hash) that are supposed to automatically interact with std::, and your intuition is correct for why it's allowed -- it only works when interacting with the namespace you are defining.  (Usually, that is; you can have some strange interactions crop up due to implicit conversions).

Fundamentally, this was a poor design decision on the part of std::unordered_set, and I do expect it to be fixed in one of the next couple of C++ updates, but alas not yet.  (Mostly because there's not an obvious fix; all of the other solutions do technically have drawbacks even if they are clearly better than the current solution -- they just want to limit it to one more fix and be done with it).

File: source/tools/MatchBin.h (hunk 1 above)
Comment (mercere99) [reply source/tools/MatchBin.h]: "I assume that the people who made that design decision for std::hash stuff know best"

Heh.  Definitely not always a good assumption.  They usually have a good reason for what they did do, but haven't always fully considered all of the drawbacks.  There are LOTS of minor examples, and a small number of majorly poor decisions (like specializing `std::vector<bool>`)

File: source/tools/MatchBin.h [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+  };
+  
+  template<size_t width>
+    struct DowningInteger{
+      double operator()(const emp::BitSet<width>& a, const emp::BitSet<width>& b){
+        emp::BitSet<width> bitDifference = ( a > b ? a - b : b - a);
+        unsigned int fields = bitDifference.GetFields();
+        double result = 0;
+        for (unsigned int i = 0; i < fields; ++i){
+          result += bitDifference.GetUInt(i) * pow(2, 32 * i);
```
[the remark this answers is not in the document]
Comment (mmore500) [reply source/tools/MatchBin.h]: Let’s just use the GetUint method of BitSet. We should just static assert that width < 32 to use this method. Let’s nix the GetFields method on BitSet so we don’t expose the underlying implementation so it is flexible to change in the future.

File: source/tools/MatchBin.h [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+    };
+  /// Selector for MatchBin stored in the struct so we cn template on it
+  /// returns sorted matches that are inside the given threshold 
+  template<typename ratio>
+  struct ThreshSelector{
+    emp::vector<size_t> operator()(emp::vector<size_t>& uids, std::unordered_map<size_t, double>& scores, size_t n){
+        
+      double thresh = ratio::num / ratio::den;
+      unsigned int i = 0;
+      if (n < log2(uids.size())){
```
Comment (mmore500) [inline source/tools/MatchBin.h]: Where are log2 and pow coming from?? We didn’t #include “math.h” in this file I wonder if it’s somehow getting included via catch. Let’s #include <cmath> and then prefix these with std::

**Comment by mmore500** [comment]
API tweak... some Selector structs use Random objects & we want to give the user control over what Random object is being used at runtime. We should have the MatchBin constructor take an argument for a emp::Ptr<Random>=nullptr (defaulted) and then pass that to the constructor for the Selector object. With some templating cleverness we might be able to just ignore the Random arg for Selector objects that don’t need Random object in the constructor.

**Comment by mmore500** [comment]
some other minor TODOs are commented in MatchBin, do a quick search thru to tackle the todos that are actionable

**Comment by mmore500** [comment]
@mercere99 I had to do some manual merging of Emphatic-related code (god knows why) in this commit [https://github.com/devosoft/Empirical/pull/240/commits/022192a3c33c37ffd332c15069af65d8d19e5dc5](https://github.com/devosoft/Empirical/pull/240/commits/022192a3c33c37ffd332c15069af65d8d19e5dc5), so it would be a good idea to double check I made the right decisions and I'm not reverting any Emphatic-related work you've done before merging this into master

**Comment by mmore500** [comment]
Otherwise, this should be ready to pull in!

**Comment by mmore500** [comment]
@mercere99 should we consider moving the Selector classes and Metric classes to a separate header file?

**Comment by mmore500** [comment]
There are somehow a bunch of regressive/unrelated diffs in here I need to go through and vacuum out. Not sure how that happened.

**Comment by mmore500** [comment]
@mercere99 we decided that caching was a critical feature for the integration of this with SignalGP so we're going to pause merging in this PR until we implement that

**Comment by mmore500** [comment]
Our first naive idea:
```
Caching chat redux:
* user should be able to disable caching
   * cache_on field inside MatchBin 
* we will have to handle caching differently based on Selector (e.g., RouletteSelector vs. RankedSelector)
   * static constexpr field cache_owner (True = MatchBin shouldn't cache, False = MatchBin should handle caching)

cache implementation inside MatchBin
* unordered_map of queries to vectors of UIDs
* if the vector isn't long enough for requested n, do another look up and store the longer lookup

clear cache function
* if Selector cache_owner = True then Matchbin forwards call ClearCache to Selector
* should ClearCache be defined in Selectors that don't own the cache (yes, add a virtual ClearCache function inside Selector base class with an emp_assert in it to make sure that if ClearCache is called it has been overridden)

things that should clear cache:
* Put
* Delete
* Clear
* GetTag should return a const reference any more, instead make a function SetTag which will clear the cache
* AdjRegulator
* SetRegulator
* the gimme... ClearCache

inside RouletteSelector:
* cache the set-up IndexMap in a unordered_map of uids to IndexMap ... ??????? can't figure this out
```

**Comment by mmore500** [comment]
Our new clever idea:

`Selector::CacheStateType` (a data type) and `Selector::InterpretCache(const CacheStateType & cache, size_t n)` (a function)
* for `RouletteSelector`
  * `Selector::CacheStateType` is `emp::IndexMap`
  * `Selector::InterpretCache(const CacheStateType & cache, size_t n)` does random look ups on `emp::IndexMap` to return `n` `uid_t` results
* for `RankedSelector`
  * `Selector::CacheStateType` is `emp::vector<uid_t>`
 * `Selector::CacheStateInterpreter(const CacheStateType & cache, size_t n)` returns the first `n` of `emp::vector<uid_t>`

in addition to  `operator()` returning `emp::vector<size_t>` each Selector will provide a `MakeCache()` function that returns `Selector::CacheStateType`
* `operator()` can internally call `MakeCache()` followed by `InterpretCache()`

user should be to disable MatchBin caching
* `MatchBin::DisableCaching()` sets internal private field `cache_on` to `false`

private `cache` member inside `MatchBin`: `std::unordered_map<Tag, Selector::CacheStateType>`

things that should clear cache:
* Put
* Delete
* Clear
* GetTag should return a const reference any more, instead make a function SetTag which will clear the cache
* AdjRegulator
* SetRegulator
* the gimme... ClearCache

give the `Selector` some kind of way (callback lambda) to call ClearCache on the MatchBin it lives inside of?
* better idea? have a `cache_ok` field that is defined in `Selector`s that need to be able to nuke the `MatcBin`'s cache... inside `MatchBin` have an `if constexpr` that checks if the `cache_ok` field exists within the Selector... and if the field exists then check that it's `true` before every access
* sfinae?

**Comment by mmore500** [comment]
Break MatchBin regulator into its own module

**Comment by mmore500** [comment]
failure counts with std random backing
```
GetUInt64 3
RandFill 2
GetInt 0
GetUInt 4
```

**Comment by mmore500** [comment]
failure counts with bad RandFill
```
GetUInt64 5
RandFill 4
GetInt 0
GetUInt 1
```

**Comment by mmore500** [comment]
failure counts with better RandFill and custom backing
```
GetUInt64 2
RandFill 0
GetInt 0
GetUInt 3
```

**Comment by mmore500** [comment]
custom backing is about twice as fast as std random backing

**Comment by mmore500** [comment]
@mercere99 I reviewed with Alex this afternoon, so it's ready for a final review from you! Perhaps we can discuss at our weekly meeting Monday?

CHANGE SUMMARY:
* add Cereal dependency to third-party
* Evolve/World.h, make Grid's `fun_get_neighbor` not return self (this can segfault systematics)
* Add local tournament select (competitors are neighbors) to WorldSelect.h
* Data: add median/percentile functionality
* EventDrivenGP: plug in MatchBin
* BitSet: rotate, mutate, fix undefined behavior
* Random: randfill
* keyname_utils: C++ side implementation of https://pypi.org/project/keyname/
* MatchBin with modular Selector, Regulator, Metrics ... and built in caching
* web/DocuExtras: manipulate a html element's styling/attributes without nuking the content inside (e.g., if you attach to an existing div in an html document and don't want to overwrite the existing content in it)
* web/Element: a div, but you choose the html tag (p, footer, header, etc.)
* web/Input: fix headaches for checkboxes and radio buttons

**Comment by mercere99** [comment]
Sounds good.
